### PR TITLE
feat: enhance advantages tracking and normalization stability in GRPO

### LINKED ISSUE #1395: Track advantages on the RL algorithms
Tracking advantages can be helpful for debugging unstable runs, we could consider tracking min, max and mean values 

Image

### PR DESCRIPTION
**Pull request by ffrujeri**
# What does this PR do ?

**Improves advantages tracking and normalization in GRPO algorithms for better training stability and debugging.**

# Issues
- Closes #1395 - Track advantages on the RL algorithms
- Related to #1292 - Investigate if TP affects Acemath recipe

# Key Changes

## 1. Enhanced Advantage Tracking
Added metrics tracking for `advantages/mean`, `advantages/max`, and `advantages/min`
- These metrics help with debugging unstable training runs by providing visibility into advantage distribution

## 2. Improved Advantage Normalization
- **New function**: `normalize_advantages_with_epsilon()` that uses epsilon (1e-6) to avoid division by zero instead of masking
- **Replaces**: Previous zero-standard-deviation masking approach in both sync and async GRPO
- **Benefits**: More numerically stable and avoids potential issues with masked advantages

## 3. Enhanced Standard Deviation Calculation
- **Fixed**: Potential numerical issues in `calculate_baseline_and_std_per_prompt()` 
- **Added**: More robust standard deviation computation with proper handling of edge cases

# Usage

The advantage tracking metrics will automatically appear in your training logs when using GRPO. For example:

```python
# These metrics will now be logged during training:
metrics = {
    "advantages/mean": 0.125,    # Mean advantage value
    "advantages/max": 2.456,     # Maximum advantage value  
    "advantages/min": -1.234,    # Minimum advantage value
    # ... other metrics
}
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ x] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ x] Did you write any new necessary tests?
- [ x] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ x] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
- **Debugging**: The new advantages metrics are particularly helpful for debugging unstable runs in math training recipes
- **Numerical Stability**: The epsilon-based normalization prevents division by zero errors that could occur with the previous masking approach
- **Backward Compatibility**: All changes are backward compatible and don't affect existing configurations
- **Performance**: No significant performance impact expected from the additional metric tracking

This improvement addresses stability issues that could affect tensor parallelism configurations in math training recipes, providing better observability and more robust normalization.

image

The comparison between baseline and treatment metrics can be seen in [this report](https://wandb.ai/nvidia/nemo-rl-acemath-rl/reports/Nemo-RL-vs-VeRL-Acemath--VmlldzoxNDU0NDg5MA)

In terms of perf time_per_step:
image

In terms of training_rewards:
image

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved numerical stability in advantage normalization during training by introducing epsilon-based handling to prevent division-by-zero errors in edge cases.

* **Tests**
  * Added comprehensive test coverage for advantage normalization functionality and baseline/standard deviation computations across various scenarios and tensor shapes.

### CODE REVIEW

File: nemo_rl/algorithms/utils.py [hunk 1]
```diff
@@ -115,6 +116,14 @@ def calculate_baseline_and_std_per_prompt(
...
+            std[prompt_idx] = (
+                (
+                    (prompt_baseline_square - prompt_baseline.square())
+                    * (num_valid / (num_valid - 1))
+                )
+                .sqrt()
+                .nan_to_num(0)
+            )
 
     std = (sq_baseline - baseline.square()).sqrt().nan_to_num(0)
```
[the remark this answers is not in the document]
Comment (ffrujeri) [reply nemo_rl/algorithms/utils.py]: Done.

**Review by terrykong — DISMISSED** [review]

**Review by terrykong — APPROVED** [review]

**Review by terrykong — APPROVED** [review]
